Synthesize buttons for embedders

dkwingsmt opened the pull request:
## Description

This PR is a placeholder for button support for embedders. Currently the embedder API does not have button information at all, which leads to `PointerData.buttons = 0` and blocks the corresponding framework change. This PR synthesizes a `buttons = kPrimaryButton` for events of down and move.

In the future when we have added `FlutterPointerEvent::buttons` (without breaking existing binaries), make sure to fix these 2 TODOs.

## Related issues
- Discussion: https://github.com/flutter/flutter/issues/31884
- Fixes: https://github.com/flutter/flutter/issues/32052
- Unblocks: https://github.com/flutter/flutter/pull/31935

stuartmorgan commented on lib/ui/window/pointer_data.h:
```diff
@@ -9,6 +9,9 @@
 namespace flutter {
 
+// Must match the button constants in events.dart.
+const int64_t kPrimaryButton = 0x01;
```
Why is this a free constant rather than an `enum class` like the others below? In my experience it's standard in C++ to use enums for flags.

It would also be good to just go ahead and add the other mouse button constants from events.dart at the same time so the list here matches, rather than being a subset.

dkwingsmt replied:
I'm not sure if it's ok to use enum as bitfield.

The reason I chose not to add other button constants is that only `kPrimaryButton` is used in code, since only it is used synthesis. When other buttons are used they will be directly copied.

I'll fix them if they're ok.

stuartmorgan replied:
> I'm not sure if it's ok to use enum as bitfield.

See https://github.com/flutter/engine/blob/master/lib/ui/semantics/semantics_node.h#L21 and later in the same file for example.

> only kPrimaryButton is used in code

We'll need all of them when it becomes part of the embedding API, since we'll need to translate each public constant to the corresponding internal constant.

dkwingsmt replied:
Note: Stuart and I discussed privately about whether to use enum types or int64_t. 

Pros of enum class:
- Type safe, such as preventing `kMousePrimaryButton | kStylusSecondaryButton`.
- Grouping comes more natively than explicit namespace.

Cons:
- Require to define `operator&` and `operator|`. It is magnified if we're to define all 4 sets of constants.
- Combined values no longer fall into enum options, which might be confusing.

The cost to implement all operators is considerable, while the benefit seems insignificant because they're not currently used in engine (also it is allowed in framework), I'm going forward with int64_t constants.

stuartmorgan commented on shell/platform/embedder/embedder.cc:
```diff
@@ -712,6 +712,25 @@ inline flutter::PointerData::SignalKind ToPointerDataSignalKind(
+    int64_t buttons,
+    flutter::PointerData::Change change) {
+  // `buttons` should be non-zero when and only when the pointer is down.
+  if (change == flutter::PointerData::Change::kDown ||
```
I would recommend a `switch` rather than an `if` here so that if a new change type is added, the compiler with warn about the need to evaluate this code to decide how to handle it (since it may be down).

dkwingsmt replied:
Do you mean not using `default` but writing all options instead?

stuartmorgan replied:
Yes.

stuartmorgan commented on shell/platform/embedder/embedder.cc:
```diff
@@ -712,6 +712,25 @@ inline flutter::PointerData::SignalKind ToPointerDataSignalKind(
+inline int64_t ToPointerDataButtons(
+    int64_t buttons,
+    flutter::PointerData::Change change) {
+  // `buttons` should be non-zero when and only when the pointer is down.
```
Is the "and only when" actually true? For instance, is sending a `cancel` with a button set actually an error? I can think of cases where I would expect that to be the event created by the platform on desktop, unless it were explicitly worked around at the platform level.

dkwingsmt replied:
Cancel is kind of special being the state of "termination". I can add it to the list (down, move or cancel), but other events should really not involve buttons.

For other events, there is a set of restrictions we hope the list of `PointerData` should obey, and the engine should regularize them before sending to framework. It is currently handled by [`PointerEventConverter `](https://github.com/flutter/flutter/blob/7beb09e76af70b92a6c03879fb22eed1ab86a8c3/packages/flutter/lib/src/gestures/converter.dart#L103). There was [a plan](https://github.com/flutter/engine/pull/8064) to move them to the engine (leaving framework with only 1-to-1 converting) but was cancelled. I'm not sure we're still going this way though.

stuartmorgan commented on shell/platform/embedder/embedder.cc:
```diff
@@ -712,6 +712,25 @@ inline flutter::PointerData::SignalKind ToPointerDataSignalKind(
+      change == flutter::PointerData::Change::kMove) {
+    if (buttons == 0) {
+      // Synthesize primary button if the embedder fails to follow the contract.
+      // TODO(tongmu): Log a warning to inform the embedder to send the correct
```
Could you change this from your username to a reference to the bug number for the bug where we were discussing this, since it provides more context on the circumstances where a warning should happen? As written here, without that context, it could encourage someone to add logging in cases where I don't think it's warranted (per my bug comments)

stuartmorgan commented on shell/platform/embedder/embedder.cc:
```diff
@@ -731,6 +750,10 @@ FlutterEngineResult FlutterEngineSendPointerEvent(
+    // TODO(tong): Change 0 to a SAFE_ACCESS to current.buttons once this
+    // field is added.
+    // See https://github.com/flutter/flutter/issues/32052#issuecomment-489278965
+    pointer_data.buttons = ToPointerDataButtons(0, pointer_data.change);
```
Since the button field will be at the end, let's put this line at the end as well; having them in a different order here seems like it introduces potential confusion.

dkwingsmt commented:
@stuartmorgan I fixed all issues you raised. Notes:
- I still decided it's better to use int64_t for button constants.
- I changed `ToPointerDataButtons` so that they no longer force `buttons = 0` for the events that are supposed to, but they should still log a warning in case of violation.

stuartmorgan commented on shell/platform/embedder/embedder.cc:
```diff
@@ -712,6 +712,42 @@ inline flutter::PointerData::SignalKind ToPointerDataSignalKind(
+  switch (change) {
+    case flutter::PointerData::Change::kDown:
+    case flutter::PointerData::Change::kMove:
+    case flutter::PointerData::Change::kCancel:
```
I don't think `kCancel` should always have non-zero buttons; it seems like it should be able to go either way. I'd expect it to be a no-op third case in the switch, with a comment explaining why.

dkwingsmt replied:
It's an oversight. Will fix.

stuartmorgan commented on shell/platform/embedder/embedder.cc:
```diff
@@ -712,6 +712,42 @@ inline flutter::PointerData::SignalKind ToPointerDataSignalKind(
+    case flutter::PointerData::Change::kHover:
+    case flutter::PointerData::Change::kUp:
+      // These kinds of change should always have a zero `buttons`. However we
+      // are not synthesizing a 0 until a solid reason is found.
```
I think we should remove this comment; I don't think we should ever plan to fix up code where embedders are explicitly doing the wrong thing. Code to see buttons is all only here for backwards compatibility with embedders that never had to send buttons before, not to error-proof incorrect implementations of embedders.

dkwingsmt replied:
Got it.

stuartmorgan commented on lib/ui/window/pointer_data.h:
```diff
@@ -9,6 +9,29 @@
 namespace flutter {
 
+// Must match the button constants in events.dart.
+namespace PointerButton {
```
Namespaces are not generally used for constants this way (I should have mentioned this in our discussion). If you're not using enum classes, name them with prefixes instead:
`kMouseButtonPrimary`, `kMouseButtonSecondary`, `kStylusContact`, `kStylusPrimary`, etc.

dkwingsmt replied:
Fixed using `enum` and fixed names, thanks.

stuartmorgan commented on lib/ui/window/pointer_data.h:
```diff
@@ -9,6 +9,29 @@
+// Must match the button constants in events.dart.
+namespace PointerButton {
+const int64_t kPrimary = 1 << 0;
+const int64_t kSecondary = 1 << 1;
```
Are the raw values in event.dart actually used? I wouldn't expose them unless there's a need; it seemed like they were more for defining the device-specific constants.

Having them here encourages confusing usage (e.g., your use of `PointerButton::kPrimary` rather than `PointerButton::kMousePrimary` in this PR even though it's a mouse), and especially so given that they don't line up with the stylus names.

dkwingsmt replied:
Yes. In fact, the general series of buttons are used more often, because most widgets are defined for cross-device behaviors, and by grouping the same "cross-device behavior" into the same value, more often than not widgets do not need to check the device kind. For example, primary mouse button, stylus contact, and touch contact are supposed to behave the same for most cases, so they share the same value `0x01`, while secondary mouse button and primary stylus button share the same value `0x02`. In this way, a button click will respond to `kPrimaryButton`, instead of `kPrimaryMouseButton` or `kStylusContact`, while a context menu will respond to `kSecondaryButton`.

stuartmorgan replied:
I see, makes sense.

I still think we shouldn't expose them in the C++ code absent a reason though; we're creating device-specific data structs, rather than consuming data in a device-agnostic way. For creating them, using the device-specific values makes much more sense to me.

dkwingsmt replied:
I agree. However I still need a value when synthesizing the primary button. Currently I'm using `kPrimaryMouseButton` but it's a little awkward. I wonder if you think it's acceptable. https://github.com/flutter/engine/pull/8873/files#diff-3784eecbfe20523d62f0a4c19cd363ebR730

stuartmorgan replied:
Why is that awkward? The device type is hard-coded to mouse, so the primary mouse button seems like the right constant to me.

stuartmorgan commented on lib/ui/window/pointer_data.h:
```diff
@@ -9,6 +9,29 @@
 
+// Must match the button constants in events.dart.
+namespace PointerButton {
+const int64_t kPrimary = 1 << 0;
```
If you're not using enum, you need to put the definitions in a cc file and declare them extern here (until we have C++17's `inline constexpr`).

You could use `enum` (not `enum class`) as we do in embedder.h to avoid the need for that (I think it'll implicitly convert to int64_t despite being smaller intrinsically, in which case it wouldn't have impacts on usage).

dkwingsmt replied:
I fixed it using `enum`.

dkwingsmt commented:
@stuartmorgan I fixed the requested changes. However,
- I'm using the "no default" strategy you suggested, I'm sure I listed all 7 options but still got `warning C4715: 'ToPointerDataButtons': not all control paths return a value`.
- https://github.com/flutter/engine/pull/8873#discussion_r282713695

stuartmorgan commented on lib/ui/window/pointer_data.h:
```diff
@@ -9,6 +9,21 @@
 namespace flutter {
 
+// Must match the button constants in events.dart.
+enum PointerButton : int64_t {
```
Why is this one enum instead of an enum for each pointer type?

dkwingsmt replied:
Fixed

stuartmorgan commented on lib/ui/window/pointer_data.h:
```diff
@@ -9,6 +9,21 @@
+  kSecondaryMouseButton = 1 << 1,
+  kMiddleMouseButton = 1 << 2,
+  kBackMouseButton = 1 << 3,
+  kForwardMouseButton = 1 << 4,
```
Since you're not using enum classes, it's very helpful to have a common prefix for the things that are conceptually grouped. At a minimum I would change these to:
```
kMousePrimaryButton
kMouseSecondaryButton
...
kStylusContact
kStylusPrimaryButton
```
etc.

I'd suggest going further to make it clear they are part of pointer button though, and do:
```
kPointerButtonMousePrimary
kPointerButtonMouseSecondary
...
kPointerButtonStylusContact
kPointerButtonStylusPrimary
```

dkwingsmt replied:
Fixed using the 2nd design
